### ISSUE TITLE
Differing behaviour when object is assigned versus when it is written inline

### ISSUE THREAD

**Issue by cdimitroulas** [Microsoft/TypeScript#30486]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:**
object inline behaviour

**Code**

```ts
type Prop = <K extends string>(path: K) => <V>(obj: { [k in K]: V }) => V;

const prop: Prop = path => obj => obj[path];

const myObj = { x: true, value: "hi" }

const a = prop('value')(myObj) // this is ok
const b = prop('value')({ x: true, value: "hi" }) // this is a compile error?
.
```

**Expected behavior:**
The prop function in the code example above only accepts an object with one property which is the key specified by the `path` argument.
It is expected that both of the following statements (assignments to `a` and `b`) above should throw a compiler error.

**Actual behavior:**
Only the statement which is assigne to `b` (where the object is written in-line) throws a compiler error

**Playground Link:** 
https://www.typescriptlang.org/play/index.html#src=type%20Prop%20%3D%20%3CK%20extends%20string%3E(path%3A%20K)%20%3D%3E%20%3CV%3E(obj%3A%20%7B%20%5Bk%20in%20K%5D%3A%20V%20%7D)%20%3D%3E%20V%3B%0D%0Aconst%20prop%3A%20Prop%20%3D%20path%20%3D%3E%20obj%20%3D%3E%20obj%5Bpath%5D%3B%0D%0A%0D%0Aconst%20myObj%20%3D%20%7B%20x%3A%20true%2C%20value%3A%20%22hi%22%20%7D%0D%0Aconst%20a%20%3D%20prop('value')(myObj)%20%2F%2F%20this%20is%20ok%0D%0Aconst%20b%20%3D%20prop('value')(%7B%20x%3A%20true%2C%20value%3A%20%22hi%22%20%7D)%20%2F%2F%20this%20is%20a%20compile%20error%3F

**Related Issues:** 
N/A

**Comment by dragomirtitian** [comment]
This is by design. [Excess property checks](https://github.com/Microsoft/TypeScript/pull/3823) kick in when you assign an object literal to a parameter that has a specific type. In this case `obj` is typed as an object that has the keys in `K`, so excess property checks will make sure only those properties are defined. If you create the object literal first then the check does not kick in and the assignment can happen. 

If you want to allow extra properties even in object literals passed in (although the check was added because this is a common error) you can type `obj` as something that extends a type that has the properties `K`:

```ts
type Prop = <K extends string>(path: K) => <T extends Record<K, any>>(obj: T) => T[K];
const prop: Prop = path => obj => obj[path];

const myObj = { x: true, value: "hi" }
const a = prop('value')(myObj) // this is ok
const b = prop('value')({ x: true, value: "hi" }) // also ok
```

**Comment by cdimitroulas** [comment]
OK thanks for the explanation and a possible solution @dragomirtitian 

I ended up with the following type signature to solve my problem which gives slightly nicer compiler errors
```ts
type Prop = <K extends string>(
  path: K
) => <V>(obj: Record<K, V> & Record<string, any>) => V;
```

Feel free to close this as this is obviously not a bug
